src/language/interpreter_new.rs

```diff
@@ -1019,316 +1019,316 @@ impl Sqrt for i64 {
     }
 }
 
+extern crate test;
+
 #[cfg(test)]
 mod tests {
+
     use super::*;
     use approx::AbsDiffEq;
     use ndarray::array;
     use std::str::FromStr;
+    use test::Bencher;
+
+    /// Creates an benchmark test for interpreter
+    /// The test does the following:
+    ///  1. Parses $glenside_str as glenside expression
+    ///  2. Creates a new Environment from the vector of (key, value) pairs in base_env
+    ///  3. Calls check_correct with the glenside expression + environment
+    /// $test_name: the name of the created benchmark test
+    /// $glenside_str: A string containing the Glenside program
+    /// $base_env: A vector or 2-tuples of key, value pairs to put into the environment
+    /// $check_correct: A closure with arguments (glenside_expr, env) that should call the interpreter
+    ///     and checks for correctness
+    macro_rules! benchmark_test {
+        ($test_name:ident, $glenside_str: expr, $base_env: expr, $check_correct: expr) => {
+            #[bench]
+            fn $test_name(b: &mut Bencher) {
+                let test = test_function!($glenside_str, $base_env, $check_correct);
+                b.iter(|| test());
+            }
+        };
+    }
+    macro_rules! test_function {
+        ($glenside_str: expr, $base_env: expr, $check_correct: expr) => {
+            || {
+                let mut env = Environment::new();
+                for (key, value) in $base_env.into_iter() {
+                    env.insert(key, value);
+                }
 
-    #[test]
-    fn compute_elementwise_add_0() {
-        let mut env = Environment::new();
-        env.insert(
-            "t",
-            array![[[1, -2], [3, 0]], [[-5, 6], [0, 8]], [[-9, 10], [11, 12]],].into_dyn(),
-        );
+                let expr = RecExpr::<Language>::from_str($glenside_str).unwrap();
 
-        let expr = RecExpr::<Language>::from_str(
-            "(compute elementwise-add
-              (access (access-tensor t) 0)
-             )",
-        )
-        .unwrap();
+                $check_correct(expr, env);
+            }
+        };
+    }
 
-        match interpret(&expr, expr.as_ref().len() - 1, &env) {
-            Value::Access(Access {
-                tensor,
-                access_axis,
-            }) => {
-                assert_eq!(access_axis, 0);
-                assert_eq!(
+    benchmark_test!(
+        compute_elementwise_add_0,
+        "(compute elementwise-add
+        (access (access-tensor t) 0)
+        )",
+        vec![(
+            "t",
+            array![[[1, -2], [3, 0]], [[-5, 6], [0, 8]], [[-9, 10], [11, 12]],].into_dyn(),
+        )],
+        |expr, env| {
+            match interpret(&expr, expr.as_ref().len() - 1, &env) {
```

Comment:
Sorry if I wasn't clear about this in my Slack message, but I was thinking the macro should call `interpret`, and then pass the result into the `check` function. So this closure would take one parameter, `value`, and then would do something like 
```rust
match value {
...what you already have
}
```
Very minor thing, but definitely makes life easier for test writers!

Reply:
oh ok, sorry the misunderstanding! changed!